LibDiff+diff+headless-browser: Support unfied hunks with context

shannonbooth opened the pull request:
The existing hunk data structure does not contain any way to easily
store information about context surrounding the additions and removals
in a hunk. While this does work fine for normal diffs (where there is
never any surrounding context) this data structure is quite limiting for
other use cases.
    
Without support for surrounding context it is not possible to:
 * Add support for unified or context format to the diff utility to
    output surrounding context.
 * Be able to implement a patch utility that uses the surrounding context
    to reliably locate where to apply a patch when a hunk range does not
    apply perfectly.
    
This patch changes Diff::Hunk such that its data structure more closely
resembles a unified diff. Each line in a hunk is now either a change,
removal, addition or context. (unlocking the above bullet points to be
implemented!)

Allowing hunks to have context inside of them exposes that HackStudio
heavily relies on there being no context in the hunks that it uses for
its' git gutter implementation. The fix here is simple - ask git to
produce us a diff that has no context in it!

Note that while testing, I also ran into https://github.com/SerenityOS/serenity/issues/19662
I've verified that this issue exists previous to any of my changes. Also checked
that after working around that issue (by removing that VERIFY_NOT_REACHED)
all of the diff offsets are calculated correctly, and shown on the right line.

Making this change allows us to implement optionally adding a specified number
of context lines when calculating the diff. With this data structure, it then becomes
trivial to output diffs in unified format with surrounding context information:

![image](https://github.com/SerenityOS/serenity/assets/35911232/e6c8fe55-f4b8-41d6-886f-a076cd7ed52b)

ADKaster commented on Userland/Libraries/LibDiff/Hunks.h:
```diff
@@ -12,25 +13,24 @@
+};
+
+struct Line {
+    char operation; // Addition '+', Removal '-', Context ' ', or Change '!'.
```
This feels like we should have an `enum Operation` here, and a `constexpr u32 symbolForOperation(Operation)` function

shannonbooth replied:
Sounds good, will do!

shannonbooth replied:
Done!

shannonbooth commented:
woo, just finished off getting diff working with these changes to produce unified diffs with context :^)

shannonbooth commented:
I was originally thinking it would be easier to get this in by splitting it up into two MRs - but perhaps it is better to include them all together in one. I've rebased my changes on top of latest master, and added support for unified diffs with context in this change set.

ADKaster reviewed (approved):
Looks cool! glad to see the new feature used in the same PR :)
